File: Data Structures & Algorithms/is-anagram/submission-1.py
# Given: two strings s and t
# return: true if the two strings are anagrams of each other(, otherwise return false.)
# Definition: An anagram is a "string" that contains the "exact same characters" as another string, but the "order" of the characters can be different.
# T/S: aim for O(n + m) time and O(1) space
# Constraints: s and t consist of lowercase English letters.
# Solution: (1)Sorting (2)Hash Map

class Solution:
    ## (1)Sorting
    # def isAnagram(self, s:str, t:str) -> bool:
    #     if len(s) != len(t):
    #         return False    #若長度不同，則不用花費排序的時間

    #     return sorted(s) == sorted(t)
    ## (2)Hash Map
    def isAnagram(self, s:str, t:str) -> bool:
        if len(s) != len(t):
            return False
        countS, countT = {}, {}
        for i in range(len(s)):
            countS[s[i]] = 1 + countS.get(s[i], 0)
            countT[t[i]] = 1 + countT.get(t[i], 0)
        return countS == countT

# def main():
#     sol = Solution()
#     while True:
#         try:
#             s, t = input("請輸入兩組英文字串(用空格間隔): ").split()
#             if not s.isalpha() or not t.isalpha():
#                 print("請輸入有效的~英文~字串。")
#             else:
#                 print(sol.isAnagram(s, t))
#         except KeyboardInterrupt:
#             print("\nExiting the program.")
#             break
#         except ValueError as e:
#             print(f"An error occurred: {e}.\n(🔥請照規則輸入!!!)")
#             continue
#         except Exception as e:
#             print(f"An error occurred: {e}.\n(🔥請照規則輸入!!!)")
#             continue

# if __name__ == '__main__':
#     main()

File: Data Structures & Algorithms/anagram-groups/submission-1.cpp
// # Given: an array of strings strs
// # Goal: group all anagrams together into sublists
// # return: the output in any order

// # def: An anagram is a string that contains the exact same characters as another string, but the order of the characters can be different.
// # Constraints: strs[i] is made up of lowercase English letters.
// # aim for a solution: with O(m * n) time and O(m) space, where m is the number of strings and n is the length of the longest string.
// # ------------------------------------------------
// # Solution: (1)Sorting(key=sorted string, value:original strings) (2)Dicationary(26字母 + count)- {tuple(count): List[str]}

class Solution {
    // sol1️⃣: sorting
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> res_map;
        for (const auto& s : strs) {
            string sortedS = s; // 排序前複製
            sort(sortedS.begin(), sortedS.end()); //in-place 不回傳
            res_map[sortedS].push_back(s);
        }
        vector<vector<string>> res;
        for (const auto& eachPair : res_map) {
            res.push_back(eachPair.second); // 🔥不是.second()!!!
        }
        return res;
    }
};